perftest/PerfTest.hpp

```diff
@@ -0,0 +1,136 @@
+    std::string _resultsDir;
+    std::shared_ptr<PerfTestSocketHandler> _handler;
+private:
+    std::chrono::steady_clock::time_point _startTime;
+    std::chrono::steady_clock::time_point _stopTime;
+    bool _measurementStarted = false;
+    bool _measurementFinished = false;
+    std::map<std::string, std::string> _results;
+
+public:
+    PerfTest(const std::string &name, const std::string &resultsDir, const std::string &server);
+protected:
+    PerfTest(const std::string &name, const std::string &resultsDir, std::shared_ptr<PerfTestSocketHandler> handler);
+
+public:
+    bool isStarted();
+    bool isFinished();
+    void abort(const std::string &message);
+    // Output results as json
+    void printResults();
+    virtual void runTest() = 0;
+protected:
+    // Result is type string because it will be printed
+    void addResult(const std::string &name, const std::string &result);
+
+// Commands to use from tests
+protected:
+    virtual void startMeasurement();
+    virtual void stopMeasurement();
+    void connect(const std::string &filePath);
+    void loadDocument(const std::string &filePath);
+    void connectAndLoad(const std::string &filePath);
+    void waitForMessage(const std::string &substring, unsigned int timeout = PERFTEST_DEFAULT_TIMEOUT);
+    void waitForIdle(size_t timeout = PERFTEST_DEFAULT_TIMEOUT);
+    void sendMessage(const std::string &message);
+    void sleep(unsigned int millis);
+    void disconnect();
+};
+
+class CyclePerfTest : public PerfTest
+{
+private:
+    pid_t child_pid = -1;
+
+public:
+    CyclePerfTest(const std::string &name, const std::string &resultsDir, const std::string &server);
+protected:
+    CyclePerfTest(const std::string &name, const std::string &resultsDir, std::shared_ptr<PerfTestSocketHandler> handler);
+
+public:
+    void startMeasurement() override;
+    void stopMeasurement() override;
+private:
+    pid_t getCoolwsdPid();
+    std::string getStringPopen(const std::string &command);
+};
+
+class MessagePerfTest : public PerfTest
+{
+private:
+    // Shared with MessagePerfTestSocketHandler
+    std::atomic<bool> _measuring = false;
+    std::atomic<unsigned int> _messageCount = 0;
+    std::atomic<unsigned int> _messageBytes = 0;
+    std::atomic<unsigned int> _messageCountTile = 0;
+    std::atomic<unsigned int> _messageBytesTile = 0;
+
+public:
+    MessagePerfTest(const std::string &name, const std::string &resultsDir, const std::string &server);
+
+public:
+    void startMeasurement() override;
+    void stopMeasurement() override;
+};
+
+// Measure both Cycles and Messages
+// Must reimplement one class to avoid diamond inheritance
+// Reimplement MessagePerfTest because it is simpler
+// Fortunately we do not have to reimplement MessagePerfTestSocketHandler
+class CombinedPerfTest : public CyclePerfTest
+{
+private:
+    // Shared with MessagePerfTestSocketHandler
+    std::atomic<bool> _measuring = false;
+    std::atomic<unsigned int> _messageCount = 0;
```

Comment:
Seems deeply inferior to tools/Reply.hpp which we could have used - which has per message statistics, latency histograms and far more ... hmm ? why did we duplicate that?